core/node/rpc/syncv3/handler/handler.go

```diff
@@ -0,0 +1,340 @@
+		SyncOp:    SyncOp_SYNC_PONG,
+		PongNonce: nonce,
+	}); err != nil {
+		s.log.Errorw("failed to add ping message to the stream updates queue", "error", err)
+	}
+}
+
+func (s *syncStreamHandlerImpl) DebugDropStream(ctx context.Context, streamId StreamId) error {
+	select {
+	case <-s.ctx.Done():
+		return s.ctx.Err()
+	default:
+	}
+
+	if err := s.eventBus.EnqueueUnsubscribe(streamId, s); err != nil {
+		return err
+	}
+
+	return s.streamUpdates.AddMessage(&SyncStreamsResponse{
+		SyncOp:   SyncOp_SYNC_DOWN,
+		StreamId: streamId[:],
+	})
+}
+
+// OnUpdate implements the eventbus.StreamSubscriber interface.
+// It is expected to receive SyncOp_SYNC_UPDATE and SyncOp_SYNC_DOWN messages here.
+func (s *syncStreamHandlerImpl) OnUpdate(update *SyncStreamsResponse) {
+	select {
+	case <-s.ctx.Done():
+		return
+	default:
+	}
+
+	if err := s.streamUpdates.AddMessage(update); err != nil {
+		s.cancel(err)
+	}
+}
+
+// processMessage processes a single message from the stream updates queue.
+// Returns true if the processor should stop processing messages.
+func (s *syncStreamHandlerImpl) processMessage(msg *SyncStreamsResponse) bool {
+	select {
+	case <-s.ctx.Done():
+		return true
+	default:
+	}
+
+	msg.SyncId = s.syncID
+	if err := s.receiver.Send(msg); err != nil {
+		s.cancel(err)
+		return true
+	}
+
+	// Special cases depending on the message type that should be applied after sending the message.
+	if msg.GetSyncOp() == SyncOp_SYNC_CLOSE {
+		// Close the operation and return from the stream updates processor.
+		s.cancel(nil)
+		return true
+	}
+
+	return false
+}
+
+// NewRegistry creates a new instance of the Registry.
+func NewRegistry(eventBus eventbus.StreamSubscriptionManager) Registry {
+	return &syncStreamHandlerRegistryImpl{
+		handlers: make(map[string]*syncStreamHandlerImpl),
+		eventBus: eventBus,
+	}
+}
+
+// Get retrieves a sync stream handler by its sync ID.
+func (s *syncStreamHandlerRegistryImpl) Get(syncID string) (SyncStreamHandler, bool) {
+	s.handlersLock.Lock()
+	handler, ok := s.handlers[syncID]
+	s.handlersLock.Unlock()
+	return handler, ok
+}
+
+// New creates a new sync stream handler and registers it in the registry.
+func (s *syncStreamHandlerRegistryImpl) New(
+	ctx context.Context,
+	syncID string,
+	receiver Receiver,
+) (SyncStreamHandler, error) {
+	s.handlersLock.Lock()
+	defer s.handlersLock.Unlock()
+
+	if _, exists := s.handlers[syncID]; exists {
+		return nil, RiverError(Err_ALREADY_EXISTS, "sync operation with the given ID already exists")
+	}
+
+	ctx, cancel := context.WithCancelCause(ctx)
+
+	handler := &syncStreamHandlerImpl{
+		ctx:           ctx,
+		cancel:        cancel,
+		log:           logging.FromCtx(ctx).Named("syncv3.handler").With("syncID", syncID),
+		syncID:        syncID,
+		receiver:      receiver,
+		eventBus:      s.eventBus,
+		streamUpdates: dynmsgbuf.NewDynamicBuffer[*SyncStreamsResponse](),
+	}
+
+	s.handlers[syncID] = handler
```

Comment:
Handlers are never removed causing a memory leak.

Reply:
Added a logic to remove completed sync operation from both handler registry and event bus.